File: finscrape/absorbed/misc/neuberg/server/src/routes/fear-greed.ts
import { Router } from 'express';

const router = Router();

// ── Types ──

interface FngApiEntry {
  value: string;
  value_classification: string;
  timestamp: string;
}

interface FngApiResponse {
  data: FngApiEntry[];
}

interface FearGreedCurrent {
  value: number;
  classification: string;
  timestamp: string;
}

interface FearGreedHistory {
  value: number;
  classification: string;
  date: string;
}

interface FearGreedData {
  current: FearGreedCurrent;
  history: FearGreedHistory[];
}

// ── In-memory cache (15min TTL) ──

let cache: { data: FearGreedData | null; expiresAt: number } = { data: null, expiresAt: 0 };
const CACHE_TTL = 15 * 60 * 1000; // 15 minutes

// ── Route ──

// GET /api/fear-greed
router.get('/', async (_req, res) => {
  try {
    const now = Date.now();
    if (cache.data && now < cache.expiresAt) {
      return res.json(cache.data);
    }

    const response = await fetch('https://api.alternative.me/fng/?limit=30&format=json');
    if (!response.ok) {
      throw new Error(`API responded with status ${response.status}`);
    }

    const json = (await response.json()) as FngApiResponse;

    if (!json.data || json.data.length === 0) {
      throw new Error('No data returned from Fear & Greed API');
    }

    const current: FearGreedCurrent = {
      value: parseInt(json.data[0].value, 10),
      classification: json.data[0].value_classification,
      timestamp: new Date(parseInt(json.data[0].timestamp, 10) * 1000).toISOString(),
    };

    const history: FearGreedHistory[] = json.data.map((entry) => ({
      value: parseInt(entry.value, 10),
      classification: entry.value_classification,
      date: new Date(parseInt(entry.timestamp, 10) * 1000).toISOString().split('T')[0],
    }));

    const data: FearGreedData = { current, history };

    cache = { data, expiresAt: now + CACHE_TTL };
    res.json(data);
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : String(err);
    console.error('[FearGreed] Error fetching data:', message);
    if (cache.data) return res.json(cache.data);
    res.status(502).json({ error: 'Failed to fetch Fear & Greed Index' });
  }
});

export default router;

File: finscrape/absorbed/trading/ai-auto-trader-ahh/server/intel/feargreed.go
package intel

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"
)

const fearGreedURL = "https://api.alternative.me/fng/?limit=1"

// fearGreedResponse represents the API response
type fearGreedResponse struct {
	Name string `json:"name"`
	Data []struct {
		Value               string `json:"value"`
		ValueClassification string `json:"value_classification"`
		Timestamp           string `json:"timestamp"`
		TimeUntilUpdate     string `json:"time_until_update"`
	} `json:"data"`
	Metadata struct {
		Error string `json:"error"`
	} `json:"metadata"`
}

// FetchFearGreed fetches the current Fear & Greed Index
func FetchFearGreed(ctx context.Context) (*FearGreedData, error) {
	req, err := http.NewRequestWithContext(ctx, "GET", fearGreedURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch fear & greed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("fear & greed API returned status %d", resp.StatusCode)
	}

	var result fearGreedResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if len(result.Data) == 0 {
		return nil, fmt.Errorf("no fear & greed data returned")
	}

	data := result.Data[0]
	value, _ := strconv.Atoi(data.Value)
	timestamp, _ := strconv.ParseInt(data.Timestamp, 10, 64)

	return &FearGreedData{
		Value:           value,
		Classification:  data.ValueClassification,
		Timestamp:       time.Unix(timestamp, 0),
		TimeUntilUpdate: data.TimeUntilUpdate,
	}, nil
}

// FormatFearGreed formats Fear & Greed data for AI consumption
func FormatFearGreed(fg *FearGreedData) string {
	if fg == nil {
		return ""
	}

	emoji := "😐"
	switch {
	case fg.Value <= 20:
		emoji = "😱"
	case fg.Value <= 40:
		emoji = "😰"
	case fg.Value <= 60:
		emoji = "😐"
	case fg.Value <= 80:
		emoji = "😀"
	default:
		emoji = "🤑"
	}

	return fmt.Sprintf(`## Market Sentiment (Fear & Greed Index)
- Value: %d/100 %s
- Classification: %s

`, fg.Value, emoji, fg.Classification)
}
